src/autograd.cpp

```diff
@@ -215,35 +159,23 @@ unsigned int cpp_tensor_register_hook (Rcpp::XPtr<XPtrTorchTensor> self, Rcpp::F
       return out;
       LANTERN_CALLBACK_END("Unknon error in hook.", NULL)
     });
-    std::future<void*> result = task.get_future();
-    
-    {
-      std::lock_guard<std::mutex> lock(tasks_mutex);
-      tasks.push_front(std::move(task));
-    }
-    
+    std::future<void*> result_fut = task.get_future();
+
+    gTasks.schedule(std::move(task));
+
     std::future_status status;
     do {
-      status = result.wait_for(std::chrono::seconds(0));
+      status = result_fut.wait_for(std::chrono::seconds(0));
       if (status == std::future_status::timeout) {
-        
-        std::unique_lock<std::mutex> lock(backward_tasks_mutex);
-        while (!backward_tasks.empty()) {
-          auto task(std::move(backward_tasks.front()));
-          backward_tasks.pop_front();
-          
-          // unlock during the task
-          lock.unlock();
-          task();
-          lock.lock();
-        }
-        
-      } 
-    } while (status != std::future_status::ready); 
-    
-    return result.get();
+        gBackwardTasks.stopWhenEmpty();
+        gBackwardTasks.run();
+      }
+      std::this_thread::yield();
```

Comment:
From docs: Provides a hint to the implementation to reschedule the execution of threads, allowing other threads to run.

What does this mean?

Reply:
It means suggesting this thread to momentarily give up its place on the CPU and possibly allow the OS to schedule other threads waiting for CPU (e.g., calling `usleep(1)` would have the same effect, but is most likely less efficient, whereas `std::this_thread::yield()` is possibly more efficient, and also expresses the intention better).

Why do it: because without `yield()` here, the future status would be polled in a tight loop, which means other threads do not get a chance to be scheduled on the same core while this loop is running. So, other threads could be potentially starving if there is no free core aside from the current one executing this loop.